src/core/pointcloud/qgspointcloudlayerrenderer.cpp

```diff
@@ -116,11 +116,14 @@ bool QgsPointCloudLayerRenderer::render()
   mAttributes.push_back( QgsPointCloudAttribute( QStringLiteral( "X" ), QgsPointCloudAttribute::Int32 ) );
   mAttributes.push_back( QgsPointCloudAttribute( QStringLiteral( "Y" ), QgsPointCloudAttribute::Int32 ) );
 
+  if ( !context.renderContext().zRange().isInfinite() ||
+       mRenderer->drawOrder2d() == QgsPointCloudRenderer::DrawOrder::BottomToTop ||
+       mRenderer->drawOrder2d() == QgsPointCloudRenderer::DrawOrder::TopToBottom )
+    mAttributes.push_back( QgsPointCloudAttribute( QStringLiteral( "Z" ), QgsPointCloudAttribute::Int32 ) );
```

Comment:
better to use a switch here -- if we added additional draw orders in future then we'd want a warning raised here

Reply:
We only need to cater here for draw orders that require the Z value, ie bottom to top and top to bottom. Also combining it with `!context.renderContext().zRange().isInfinite() ||` will look a little ugly imho.